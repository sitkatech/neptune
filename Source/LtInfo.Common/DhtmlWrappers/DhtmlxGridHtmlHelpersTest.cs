/*-----------------------------------------------------------------------
<copyright file="DhtmlxGridHtmlHelpersTest.cs" company="Sitka Technology Group">
Copyright (c) Sitka Technology Group. All rights reserved.
<author>Sitka Technology Group</author>
</copyright>

<license>
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Affero General Public License <http://www.gnu.org/licenses/> for more details.

Source code is available upon request via <support@sitkatech.com>.
</license>
-----------------------------------------------------------------------*/
using System.Collections.Generic;
using ApprovalTests;
using ApprovalTests.Reporters;
using LtInfo.Common.ModalDialog;
using LtInfo.Common.Views;
using NUnit.Framework;

namespace LtInfo.Common.DhtmlWrappers
{
    [TestFixture]
    public class DhtmlxGridHtmlHelpersTest
    {
        protected const string TestControllerName = "TestController";

        [Test]
        [UseReporter(typeof(DiffReporter))]
        public void BuildGridColumnsTest()
        {
            const string indent = "";
            var gridSpec = new TestGridSpec();
            var result = DhtmlxGridHtmlHelpers.BuildGridColumns(gridSpec, indent);
            Approvals.Verify(result);
        }

        [Test]
        public void IsUsingSmartRenderingWithColumnsThatHaveTotalsTest()
        {
            var gridSpec = new TestGridSpec();
            Assert.That(DhtmlxGridHtmlHelpers.IsUsingSmartRendering(gridSpec), Is.False, "Should not be using smart rendering because we have a grid spec that has a total column");
        }

        [Test]
        public void IsUsingSmartRenderingWithColumnsThatHaveNoTotalsTest()
        {
            var gridSpec = new TestGridSpecWithNoTotalColumns();
            Assert.That(DhtmlxGridHtmlHelpers.IsUsingSmartRendering(gridSpec), Is.True, "Should be using smart rendering because we have a grid spec that has a total column");
        }

        [Test]
        [UseReporter(typeof(DiffReporter))]
        public void VerifyJavascriptForDhtmlxGrid()
        {
            var gridSpec = new TestGridSpec();
            const string gridName = "testGridName";


            var testGridSpecClasses = new List<TestGridSpecClass>();
            testGridSpecClasses.Add(new TestGridSpecClass(1, "One", true, 1000m));
            testGridSpecClasses.Add(new TestGridSpecClass(2, "Two", true, 2000m));
            testGridSpecClasses.Add(new TestGridSpecClass(3, "Three", false, 3000m));
            testGridSpecClasses.Add(new TestGridSpecClass(4, "Four", true, 4000m));
            testGridSpecClasses.Add(new TestGridSpecClass(5, "Five", false, 5000m));
            testGridSpecClasses.Add(new TestGridSpecClass(6, "Six", true, 6000m));

            var result = DhtmlxGridHtmlHelpers.DhtmlxGrid(gridSpec,
                                                          gridName,
                                                          string.Format("{0}/ListGridDataXml", TestControllerName),
                                                          "height:250px;");
            Approvals.Verify(result);
        }

        private class TestGridSpecClass
        {
            public readonly int PrimaryKey;
            public readonly string DisplayName;
            public readonly bool IsActive;
            public readonly decimal? Amount;

            public TestGridSpecClass(int primaryKey, string displayName, bool isActive, decimal? amount)
            {
                PrimaryKey = primaryKey;
                DisplayName = displayName;
                IsActive = isActive;
                Amount = amount;
            }
        }

        private class TestGridSpec : GridSpec<TestGridSpecClass>
        {
            public TestGridSpec()
            {
                ObjectNameSingular = "SOY";
                ObjectNamePlural = "SOYs";
                GridInstructionsWhenEmpty = "I am empty";

                // Edit SOY
                Add(string.Empty,
                    m =>
                    {
                        // Edit button
                        // -----------
                        var contentUrl = string.Format("{0}/EditAction/{1}", TestControllerName, m.PrimaryKey);
                        var dialogTitle = string.Format("Edit this {0}", m.DisplayName);
                        var dialogForm = new ModalDialogForm(contentUrl, 350, dialogTitle);
                        return DhtmlxGridHtmlHelpers.MakeEditIconAsModalDialogLinkBootstrap(dialogForm);

                    },
                    35);

                // Delete SOY
                Add(string.Empty,
                    m =>
                    {
                        var contentUrl = string.Format("{0}/DeleteAction/{1}", TestControllerName, m.PrimaryKey);
                        var deleteLink = DhtmlxGridHtmlHelpers.MakeDeleteIconAndLinkBootstrap(contentUrl, true);
                        return deleteLink;
                    },
                    35);

                Add("Display Name", m => m.DisplayName, 200);
                Add("Is Active", m => m.IsActive.ToYesNo(), 60, DhtmlxGridColumnFilterType.SelectFilterStrict);
                Add("Amount", m => m.Amount, 100, DhtmlxGridColumnFormatType.Currency, DhtmlxGridColumnAggregationType.Total);
            }
        }

        private class TestGridSpecWithNoTotalColumns : GridSpec<TestGridSpecClass>
        {
            public TestGridSpecWithNoTotalColumns()
            {
                ObjectNameSingular = "SOY";
                ObjectNamePlural = "SOYs";
                GridInstructionsWhenEmpty = "I am empty";

                // Edit SOY
                Add(string.Empty,
                    m =>
                    {
                        // Edit button
                        // -----------
                        var contentUrl = string.Format("{0}/EditAction/{1}", TestControllerName, m.PrimaryKey);
                        var dialogTitle = string.Format("Edit this {0}", m.DisplayName);
                        var dialogForm = new ModalDialogForm(contentUrl, 350, dialogTitle);
                        return DhtmlxGridHtmlHelpers.MakeEditIconAsModalDialogLinkBootstrap(dialogForm);

                    },
                    35);

                // Delete SOY
                Add(string.Empty,
                    m =>
                    {
                        var contentUrl = string.Format("{0}/DeleteAction/{1}", TestControllerName, m.PrimaryKey);
                        var deleteLink = DhtmlxGridHtmlHelpers.MakeDeleteIconAndLinkBootstrap(contentUrl, true);
                        return deleteLink;
                    },
                    35);

                Add("Display Name", m => m.DisplayName, 200);
                Add("Is Active", m => m.IsActive.ToYesNo(), 60, DhtmlxGridColumnFilterType.SelectFilterStrict);
                Add("Amount", m => m.Amount, 100, DhtmlxGridColumnFormatType.Currency);
            }
        }
    }
}
