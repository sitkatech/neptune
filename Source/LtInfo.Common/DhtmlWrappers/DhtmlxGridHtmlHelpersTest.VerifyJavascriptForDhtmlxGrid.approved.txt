
<div id ="testGridNameContainerDivID">
    <div id="testGridNameLoadingBar" style="display:none"></div>
    <div id="testGridNameMetaDivID" class="DhtmlxGridMeta"></div>
    <div id="testGridNameDivID" style="height:250px;"></div>
    <script type="text/javascript">
    // <![CDATA[
        
    jQuery(document).ready(function ()
    {
        // Initialize Grid
        Sitka.testGridName = new Sitka.Grid.Class.Grid("testGridName", "testGridNameDivID", null, "scrollToFirst", "mm", 30);
        // Initialize Grid Columns
        Sitka.testGridName.setGridColumns(
        [
            new Sitka.Grid.Class.GridColumn("Column0", "", "35", "left", "ro", "htmlstring", "#select_filter_html_strict", "noneSitka"),
            new Sitka.Grid.Class.GridColumn("Column1", "", "35", "left", "ro", "htmlstring", "#select_filter_html_strict", "noneSitka"),
            new Sitka.Grid.Class.GridColumn("DisplayName", 'Display Name', "200", "left", "rotxt", "str", "#text_filter", "noneSitka"),
            new Sitka.Grid.Class.GridColumn("IsActive", 'Is Active', "60", "left", "rotxt", "str", "#select_filter_strict", "noneSitka"),
            new Sitka.Grid.Class.GridColumn("Amount", 'Amount', "100", "right", "ron", "num", "#numeric_filter", "currencySitka")
        ]);

        var columnFilterList = "#select_filter_html_strict,#select_filter_html_strict,#text_filter,#select_filter_strict,#numeric_filter";
        Sitka.testGridName.grid.attachFooter(",,,,{#stat_total}", ["","","","","text-align:right"]);
        
        Sitka.testGridName.buildWithArguments(null, null, columnFilterList, "TestController/ListGridDataXml", false, null);

        // Show loading bar
        jQuery("#testGridNameLoadingBar").show();

        Sitka.testGridName.grid.attachEvent("onXLE", function (gridObj, count){
            var showFilterBar = true;
            Sitka.testGridName.unfilteredRowCount = Sitka.testGridName.grid.getRowsNum();
            if(showFilterBar)
            {
                Sitka.testGridName.setupCookieFiltering("testGridNameFilteredButton");
                Sitka.testGridName.setupFilterCountElement("testGridNameFilteredRowCount");
                Sitka.testGridName.setFilteringButtonTagName("testGridNameFilteredButton");
            }
            jQuery("#testGridNameFilteredRowCount").text(Sitka.testGridName.unfilteredRowCount);
            jQuery("#testGridNameUnfilteredRowCount").text(Sitka.testGridName.unfilteredRowCount);
            jQuery("#testGridNameLoadingBar").hide();
            jQuery("#testGridNameDivID").show();
            // if there are no rows, don't show grid and show a "No records available" message
            if(Sitka.testGridName.unfilteredRowCount > 0)
            {
                Sitka.testGridName.showHideFilterRow(showFilterBar);
                Sitka.testGridName.hideGridInstructions();
            }
            else
            {
                Sitka.testGridName.showHideFilterRow(false);
                Sitka.testGridName.setGridInstructions("<div style=\"padding:10px; font-weight:bold\">I am empty</div>", true);
            }
        });
        
        
    });
    // ]]>
    </script>
</div>